crates/engine/tree/src/download.rs

```diff
@@ -58,6 +61,10 @@ where
     set_buffered_blocks: BinaryHeap<Reverse<OrderedSealedBlockWithSenders>>,
     /// Engine download metrics.
     metrics: BlockDownloaderMetrics,
+    /// Set of hashes for in-flight full block requests from the previous poll
+    previous_full_block_hashes: HashSet<B256>,
+    /// Set of (`start_hash`, count) for in-flight block range requests from the previous poll
+    previous_block_range_requests: HashSet<(B256, u64)>,
```

Comment:
not sure we really need to store these, we're only interested if we receive a request that isn't already in progress which can currently happen if we don't have any peers yet but based on a fcu we issue requests for the same block multiple times (this can also be improved separately)

so we can already lookup the inflight_ requests and check if this is a new request and only then emit an event.

for example we already check for single blocks if we have the hash inflight

https://github.com/paradigmxyz/reth/blob/29dc8fc05f74111c9ff67ddcd03638ee081b1dc0/crates/engine/tree/src/download.rs#L123-L125

we can do the same for range checks as well by exposing these fields

https://github.com/paradigmxyz/reth/blob/29dc8fc05f74111c9ff67ddcd03638ee081b1dc0/crates/net/p2p/src/full_block.rs#L371-L374

in order to emit events we can add a simple VecDeque of `DownloadOutcome` where we can store  pending events for _new_ requests and then simply pop them on poll